TokenBucket.php

```diff
@@ -0,0 +1,129 @@
+<?
+
+use iFixit\Matryoshka\Backend as Backend;
+
+/**
+ * Implements the token bucket algorithm and stores tokens in cache.
+ */
+class TokenBucket {
+   // String to identify this bucket uniquely in cache.
+   private $identifier;
+   // TokenRate is when your parents try to ration your arcade spending.
+   private $rate;
+   // Where the token is to be stored.
+   private $storage;
+
+   /**
+    * @param $identifier string to uniqely store the bucket.
+    */
+   public function __construct($key, Backend $backend, TokenRate $rate, $max) {
```

Comment:
I think the token rate already sets the max. The real rate is stored in terms of `tokens per second` but since our token rate class takes in both numbers, the numerator can be used as the max.